File: scripts/generate-pyproject.toml.py
#!/usr/bin/env python
"""Update pyproject.toml from ``requirements/*.txt``."""
import os
import sys

import tomli_w

try:
    import tomllib  # type: ignore
except ImportError:
    import tomli as tomllib

with open(sys.argv[1], "rb") as f:
    data = tomllib.load(f)

requirements = {}
for dirname, j, k in os.walk("requirements"):
    for filename in k:
        requirements[filename.split(".")[0]] = {
            "file": os.path.join(dirname, filename)
        }
dynamic = {}
dynamic["optional-dependencies"] = requirements
data["tool"]["setuptools"]["dynamic"] = dynamic
print(tomli_w.dumps(data), end="")

File: scripts/generate-pyproject.toml.pl
#!/usr/bin/env -S perl -0777pi
BEGIN {
  $template = << 'EOF';
[tool.setuptools.dynamic.optional-dependencies.color]
file = "requirements/color.txt"
EOF
  while ($#ARGV > 0) {
    $name = shift;
    $name =~ s=.*requirements/(.+?)\.txt=$1=;
    $temp = $template;
    $temp =~ s/color/$name/g;
    push @temps, $temp;
  }
}
$dependencies = join "\n", @temps;
s=(# begin: scripts/generate-pyproject\.toml\.pl\n).*?(# end: scripts/generate-pyproject\.toml\.pl)=$1$dependencies$2=s;
